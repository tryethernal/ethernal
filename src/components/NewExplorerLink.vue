<template>
    <div>
        <v-row>
            <v-col cols="4" :class="{ 'pb-1': index == 0, 'py-0': index > 0 }">
                <v-text-field
                    v-model="link.url"
                    :rules="[v => !!v] || 'URL is required'"
                    small
                    variant="outlined"
                    density="compact"
                    label="URL *">
                </v-text-field>
            </v-col>
            <v-col cols="4" :class="{ 'pb-1': index == 0, 'py-0': index > 0 }">
                <v-text-field
                    v-model="link.name"
                    :rules="[v => !!v] || 'Name is required'"
                    small
                    variant="outlined"
                    density="compact"
                    label="Name *">
                </v-text-field>
            </v-col>
            <v-col cols="4" :class="{ 'pb-1': index == 0, 'py-0': index > 0 }">
                <v-autocomplete small variant="outlined" density="compact"
                    label="Icon"
                    v-model="link.icon"
                    :items="icons"
                    :loading="loading"
                    @update:search="search"
                    item-title="name"
                    :item-value="getItemValue"
                    auto-select-first
                    persistent-hint
                    hide-no-data
                    no-filter>
                    <template v-slot:selection="{ props, item }">
                        <v-icon class="mr-2 my-1" :icon="`mdi-${item.raw.name || item.raw.split('mdi-')[1]}`"></v-icon>{{ iconName }}
                    </template>
                    <template v-slot:item="{ props, item }">
                        <v-list-item v-bind="props"
                            :prepend-icon="`mdi-${item.raw.name || item.raw.split('mdi-')[1]}`"
                            :title="item.raw.name"></v-list-item>
                    </template>
                    <template v-slot:append>
                        <v-btn density="compact" variant="text" icon="mdi-delete" size="small" color="error" @click="remove()"></v-btn>
                    </template>
                </v-autocomplete>
            </v-col>
        </v-row>
    </div>
</template>
<script>

export default {
    name: 'NewExplorerLink',
    props: ['name', 'url', 'icon', 'index', 'uid'],
    data: () => ({
        link: { url: null, name: null, icon: null, uid: null },
        icons: [],
        loading: false,
    }),
    mounted() {
        this.link = {
            url: this.url,
            name: this.name,
            icon: this.icon,
            uid: this.uid
        };
        if (this.icon)
            this.icons.push({ name: this.icon.split('mdi-')[1] });
    },
    methods: {
        remove() {
            this.$emit('removed', this.uid);
        },
        getItemValue: data => `mdi-${data.name}`,
        search(query) {
            if (!query) return this.icons = [];

            this.loading = true;
            this.$server.searchIcon(query)
                .then(({ data }) => this.icons = data)
                .catch(console.error)
                .finally(() => this.loading = false);
        }
    },
    computed: {
        iconName() {
            const name = this.link.icon.split('mdi-')[1];
            if (name.length > 4)
                return name.slice(0, 4) + '...';
            return name;
        }
    },
    watch: {
        'link.url'() {
            this.$emit('updated', { link: this.link, index: this.index });
        },
        'link.name'() {
            this.$emit('updated', { link: this.link, index: this.index })
        },
        'link.icon'() {
            this.$emit('updated', { link: this.link, index: this.index })
        }
    }
}
</script>
