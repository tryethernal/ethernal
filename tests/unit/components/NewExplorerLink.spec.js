import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import NewExplorerLink from '@/components/NewExplorerLink.vue';

let helper;
beforeEach(() => {
    jest.clearAllMocks()
    helper = new MockHelper();
});

describe('NewExplorerLink.vue', () => {
    it('Should display a new link form', async () => {
        const wrapper = helper.mountFn(NewExplorerLink);
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display an existing link form', async () => {
        jest.spyOn(helper.mocks.server, 'searchIcon').mockResolvedValueOnce({ data: { name: 'twitter' }});
        const wrapper = helper.mountFn(NewExplorerLink, {
            propsData: {
                url: 'http://twitter.com',
                name: 'twitter',
                icon: 'mdi-twitter',
                uid: 1234
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
