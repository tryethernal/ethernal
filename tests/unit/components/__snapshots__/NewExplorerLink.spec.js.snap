// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewExplorerLink.vue Should display a new link form 1`] = `
<div>
  <div class="row">
    <div class="col col-4">
      <div class="v-input v-input--has-state v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined error--text">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light error--text" style="left: 0px; position: absolute;">URL *</label><input small="" id="input-4" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light error--text" role="alert">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-4">
      <div class="v-input v-input--has-state v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined error--text">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-7" class="v-label theme--light error--text" style="left: 0px; position: absolute;">Name *</label><input small="" id="input-7" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light error--text" role="alert">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-4">
      <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select v-autocomplete">
        <div class="v-input__control">
          <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-10" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Icon</label><input small="" id="input-10" type="text" autocomplete="off"><input type="hidden"></div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
        <div class="v-input__append-outer"><button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light error--text"></i></span></button></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NewExplorerLink.vue Should display an existing link form 1`] = `
<div>
  <div class="row">
    <div class="col col-4">
      <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-20" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">URL *</label><input small="" id="input-20" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-4">
      <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><label for="input-23" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Name *</label><input small="" id="input-23" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col col-4">
      <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined v-select v-autocomplete">
        <div class="v-input__control">
          <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-26" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: -24px;"><span class="notranslate">​</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-26" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Icon</label><input small="" id="input-26" type="text" autocomplete="off">
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-twitter theme--light"></i></div>
              </div><input type="hidden" value="mdi-twitter">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
        <div class="v-input__append-outer"><button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-delete theme--light error--text"></i></span></button></div>
      </div>
    </div>
  </div>
</div>
`;
